docs/reference/setup/jvm-options.asciidoc

```diff
@@ -5,14 +5,21 @@ You should rarely need to change Java Virtual Machine (JVM) options. If you do,
 the most likely change is setting the <<heap-size,heap size>>. The remainder of
 this document explains in detail how to set JVM options.
 
-The preferred method of setting JVM options (including system properties and JVM
-flags) is via the `jvm.options` configuration file. The default location of this
-file is `config/jvm.options` (when installing from the tar or zip distributions)
-and `/etc/elasticsearch/jvm.options` (when installing from the Debian or RPM
-packages).
+The preferred method of setting or overriding JVM options (including system
+properties and JVM flags) is via custom JVM options files. The root
+`jvm.options` configuration file is `config/jvm.options` and custom JVM options
+files can be added to `config/jvm.options.d/` (when installing from the tar or
+zip distributions), and `/etc/elasticsearch/jvm.options` and custom JVM options
+files can be added to `/etc/elasticsearch/jvm.options.d/` (when installing from
+the Debian or RPM packages). When using the
+<<docker, Docker distribution of {es}>> you can bind mount these files into
+`/usr/share/elasticsearch/config/jvm.options.d/`. You should never need to
+modify the root `jvm.options` file instead preferring to use custom JVM options
+files. Custom JVM options files are processed in lexicographic order (although
```

Comment:
This last sentence seems to be contradicting itself. Do we use lexicographic order, or is this non-deterministic? Lexicographic order is pretty strictly defined. If we are just iterating over a directory list (which it seems we are) then we shouldn't imply any kind of ordering guarantees. FWIW, it would be trivial for us to actually perform a sort here, which would allow folks to do stuff like `10_options` and `00_options` for layering overrides in a reliable way.

Reply:
We do sort them, see https://github.com/elastic/elasticsearch/pull/51882/files#diff-38516abb6f177c355f29699955e7e60eR72 in the implementation. However, this relies on the `Path#compareTo` implementation which explicitly says:

```
     * Compares two abstract paths lexicographically. The ordering defined by
     * this method is provider specific, and in the case of the default
     * provider, platform specific. This method does not access the file system
     * and neither file is required to exist.
```

"Provider" here refers to the filesystem provider. I am assuming (but did not investigate) this enables to account for odd sorting behavior such as on Windows via `StrCmpLogicalW` which sorts filenames by like `f-1`, `f-2`, `f-10` in the order `f-1`, `f-2`, `f-10` instead of `f-1`, `f-10`, `f-2`.

Reply:
I'm inclined to say that if this differs across providers we should even mention that it's lexicographic or not, or even perform the sort for that matter. If we do want to order, we should use a JDK implementation that isn't filesystem dependent, such as simply doing a `String` ordering using the filename. Which I think is effectively going to be what you described in your example.